Add ability to disable and enable subscription's callbacks

MichaelOrlov opened the pull request:
## Description

This PR adds a new API to be able to disable and enable subscription's callbacks.
<!--
Please include a summary of the changes and the related issue.
Highlight any key points or areas of concern.
-->

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.*
-->

- Fixes #https://github.com/ros2/rclcpp/issues/2984

### Is this user-facing behavior change?
Yes. The new API `disable_callbacks()` and  `enable_callbacks()` are available for the `SubscriptionBase` class.

### Did you use Generative AI?
Yes, I am using GitHub Copilot, GPT-5.0 in my workflow to help with Doxygen comments and unit tests.

### Additional Information
This PR can't be backported due to the ABI breaking changes. It was added two new virtual functions.

MichaelOrlov committed:
Add disable(enable)_callbacks() API to the subscriptions

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov committed:
Add test coverage for the disable(enable)_calbacks()

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov commented:
@ros-pull-request-builder retest this please

ahcorde commented on rclcpp/test/rclcpp/test_any_subscription_callback.cpp:
```diff
@@ -207,6 +207,126 @@ TEST_F(TestAnySubscriptionCallback, unset_dispatch_throw) {
+// Testing disable and enable callbacks
+//
+TEST_F(TestAnySubscriptionCallback, disable_enable_callbacks) {
+  std::atomic<int> callback_count{0};
```
include `<atomic>`

MichaelOrlov replied:
Fixed

ahcorde commented on rclcpp/test/rclcpp/test_generic_pubsub.cpp:
```diff
@@ -314,3 +315,130 @@ TEST_F(RclcppGenericNodeFixture, generic_subscription_different_callbacks)
+  const std::string type = "test_msgs/msg/Strings";
+  const auto qos = rclcpp::QoS(10U);
+
+  std::atomic<int> callback_count{0};
```
include `<atomic>`

MichaelOrlov replied:
Fixed

ahcorde commented on rclcpp/test/rclcpp/test_generic_pubsub.cpp:
```diff
@@ -314,3 +315,130 @@ TEST_F(RclcppGenericNodeFixture, generic_subscription_different_callbacks)
+  auto start = std::chrono::steady_clock::now();
+  while (callback_count.load() == 0 && std::chrono::steady_clock::now() - start < 30s) {
+    rclcpp::spin_some(node_);
+    std::this_thread::sleep_for(10ms);
```
include `<thread>`

MichaelOrlov replied:
Fixed

ahcorde commented on rclcpp/test/rclcpp/test_subscription.cpp:
```diff
@@ -668,3 +668,206 @@ TEST_P(TestSubscriptionInvalidIntraprocessQos, test_subscription_throws_intrapro
+{
+  const std::string topic_name = "~/test_disable";
+  initialize();
+  std::atomic<int> callback_count{0};
```
include `<atomic>`

MichaelOrlov replied:
Fixed

MichaelOrlov committed:
Address code review comments. Add missing includes in tests

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

ahcorde commented:
Pulls: ros2/rclcpp#2985
Gist: https://gist.githubusercontent.com/ahcorde/a411230422d2605afd32e975ecd65fad/raw/905ba8c783b44ae6c3ff1615db00916120eb0d5b/ros2.repos
BUILD args:  --packages-above-and-dependencies ‎rclcpp
TEST args:  --packages-above ‎rclcpp
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/17519
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=26023)](http://ci.ros2.org/job/ci_linux/26023/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=19829)](http://ci.ros2.org/job/ci_linux-aarch64/19829/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=5575)](http://ci.ros2.org/job/ci_linux-rhel/5575/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25966)](http://ci.ros2.org/job/ci_windows/25966/)

MichaelOrlov commented:
Interesting. The CI failed due to the 
> Package '‎rclcpp' specified with --packages-above-and-dependencies was not found

Believe me or not, I didn't delete it in this PR ;)

ahcorde commented:
Pulls: ros2/rclcpp#2985
Gist: https://gist.githubusercontent.com/ahcorde/107f28b5a40704d3cc70d5a20fe428d9/raw/905ba8c783b44ae6c3ff1615db00916120eb0d5b/ros2.repos
BUILD args:  --packages-above-and-dependencies rclcpp
TEST args:  --packages-above rclcpp
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/17524
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=26027)](http://ci.ros2.org/job/ci_linux/26027/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=19834)](http://ci.ros2.org/job/ci_linux-aarch64/19834/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=5583)](http://ci.ros2.org/job/ci_linux-rhel/5583/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25970)](http://ci.ros2.org/job/ci_windows/25970/)

ahcorde commented:
> Interesting. The CI failed due to the
> 
> > Package '‎rclcpp' specified with --packages-above-and-dependencies was not found
> 
> Believe me or not, I didn't delete it in this PR ;)

:smile: 

My terminal introduced a wierd character, relaunching again

jmachowinski commented on rclcpp/include/rclcpp/experimental/subscription_intra_process_base.hpp:
```diff
@@ -87,6 +87,16 @@ class SubscriptionIntraProcessBase : public rclcpp::Waitable
+  /// Enable the callbacks to be called
+  RCLCPP_PUBLIC
+  virtual
+  void enable_callbacks() = 0;
```
If I remember the discussions back about this, with was meant to be a terminal feature. Meaning you should only call disable shortly before shutdown / deletion. Therefore there should be no enable method.

fujitatomoya replied:
for general application use case, i think that this is okay to keep. the application disables callback temporary and then it can turn it back on via enable_callback method.

MichaelOrlov replied:
Yep. I thought, why not? 
It is a good practice to add `enable` as well as `disable` methods if it is trivial.
First of all, it can simplify many tests and increase coverage. And as @fujitatomoya already mentioned, one may want to temporarily pause the subscription and all associated event callbacks with it and then resume it.

jmachowinski commented on rclcpp/include/rclcpp/any_subscription_callback.hpp:
```diff
@@ -470,6 +495,10 @@ class AnySubscriptionCallback
     const rclcpp::MessageInfo & message_info)
   {
     TRACETOOLS_TRACEPOINT(callback_start, static_cast<const void *>(this), false);
+    if (callback_disabled_.load()) {
```
Hm I don't know about this part. If we still call the tracepoints, we will see a callback execution in the traces which is misleading. I think it wold be better to check for disabling in the subscription and not call the dispatch at all.

fujitatomoya replied:
second this approach. collecting trace metrics from empty function could mean almost nothing for user. this comment applies to else where.

MichaelOrlov replied:
Thanks for catching this. I will move the check upper before tracing point

MichaelOrlov replied:
@jmachowinski I fixed it in a new commit [Don't trace callback in dispatch methods if callbacks are disabled](https://github.com/ros2/rclcpp/pull/2985/commits/bf5f838885186ac3eca913c3c8918898afe20cce) please verify.

jmachowinski reviewed (changes requested):
Thanks for the patch, I wanted to look into this myself for ages, but just did not find any time...

fujitatomoya reviewed (commented):
overall, lgtm.

MichaelOrlov committed:
Don't trace callback in dispatch methods if callbacks are disabled

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

jmachowinski commented:
I had a second thought about the current approach, and I think it does not fix the issue, just makes it less likely. The subscriber can be deleted at any point in time, even during the execution of callback. 
I guess to solve this we will need to hold a lock during callback and we will need to acquire the lock during disabling.

MichaelOrlov commented:
> I had a second thought about the current approach, and I think it does not fix the issue, just makes it less likely. The subscriber can be deleted at any point in time, even during the execution of callback. I guess to solve this we will need to hold a lock during callback and we will need to acquire the lock during disabling.

But is it really a problem? The callback has its own liveliness time anyway. 
Even if the subscription will be deleted at the time when the callback is executing (which shouldn't be by design since the callback is called from the executor via the dispatch method and we keep an instance of the subscription there), it will not cause any faulty situations since the callback is independent from the subscription.
The key point is to make sure that no callbacks are called after we call `disable_callbacks()`.

jmachowinski commented:
> But is it really a problem? The callback has its own liveliness time anyway. 

Yes, the problem is that you can't figure out when it is safe to delete the object the callback is pointing to.

Lets say you have something like class MyMessageDisapatcher that handles all your callbacks from the subscriber.
You live in a lifecycle world, and by policy you delete all objects used in the node on cycling from running -> configured-> inactive -> active on system error recovery. The command to cycle comes from a service call, using a different callback group, therefore the callback is executed in a second thread. In this scenario (and by the way this is not made up, this is our setup) you can't figure out when it is safe to delete MyMessageDisapatcher. If the disable_callback call does not block or somehow ensures, that no callback is going on, code that calls disable_callbacks and afterwards directly deletes MyMessageDisapatcher still triggers an invalid memory access.

MichaelOrlov commented:
@jmachowinski, I see your use case. In your scenario, the problematic part is that you need to delete `MyMessageDisapatcher` with the callbacks by the service request. 
I will try to implement a lock as you described inside `disable_callbacks` and callback dispatch functions. 
However, it seems your workflow is a bit suboptimal and not robust by design. 
Maybe make sense to redesign it such a way that `MyMessageDisapatcher` is persistent and lives as long as the node is live, but deletes/recreates subscriptions by the service request, i.e., when the lifecycle node needs to change state.

MichaelOrlov committed:
Protect enable/disable callbacks and callbacks execution with mutex

Reasoning:
 To avoid possible UB when callbacks disabled during callback
 execution and callback handler object deleted while execution hasn't
 been finished yet.

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov commented:
@jmachowinski Done. Please see [Protect enable/disable callbacks and callbacks execution with mutex](https://github.com/ros2/rclcpp/pull/2985/commits/d06a07deaefd556c2af71be725b1d5b6d0ebfa7e)

jmachowinski commented on rclcpp/include/rclcpp/event_handler.hpp:
```diff
@@ -310,12 +323,28 @@ class EventHandler : public EventHandlerBase
 
   ParentHandleT parent_handle_;
   EventCallbackT event_callback_;
+  std::mutex callback_mutex_;
```
EventHandlerBase already contains a 
  std::recursive_mutex callback_mutex_;

I wonder why the compiler is not yelling that you are shadowing it...

MichaelOrlov replied:
I renamed them to avoid collisions and UB.

jmachowinski commented on rclcpp/include/rclcpp/any_subscription_callback.hpp:
```diff
@@ -972,6 +1016,8 @@ class AnySubscriptionCallback
   //   http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2162r0.html
   // For now, compose the variant into this class as a private attribute.
   typename HelperT::variant_type callback_variant_;
+  std::mutex callback_mutex_;
```
I would make the mutex recursive, someone will call it in the callbacks itself...

MichaelOrlov replied:
I would be against it. I don't like the idea of having a mutex locked for each callback execution from the first hand.
Because nothing is for free and comes with its own price and drawbacks.
Using a recursive mutex will add even more penalty on performance and open more opportunities for unexpected race conditions. 
Need to use a recursive mutex where it is really a recursion; in all other cases, need to redesign the code so as to use a regular std::mutex.

jmachowinski replied:
https://quick-bench.com/q/s8Qi6dQxBiXkczVMGxDlNxXjL64

The performance penalty is neglectable. I do understand were you are coming from, but trust me, this mutex is the last of our performance problems. I would choose a recursive mutex with a minimal performance penalty over a potential deadlock by (valid) user code.

MichaelOrlov replied:
@jmachowinski I changed to the `recursive_mutex`. 
However, my primary concern is that it will open up more opportunities for unexpected race conditions.  
Though I hope it will not be the case.

jmachowinski commented on rclcpp/include/rclcpp/subscription_base.hpp:
```diff
@@ -212,6 +212,16 @@ class SubscriptionBase : public std::enable_shared_from_this<SubscriptionBase>
   std::shared_ptr<rclcpp::SerializedMessage>
   create_serialized_message() = 0;
 
+  /// Disable callbacks from being called
```
```suggestion
  /// Disable callbacks from being called
  /**
  * This method will block, until any callback currently being  executed is finished.
  * This method is thread safe, and provides a safe way to atomically disable the callbacks
  * in a multi threaded environment. E.g. It is safe to delete the objects were the subscriber
  * is calling to after calling this method.
  */
```

jmachowinski reviewed (changes requested):
Rest looks fine.

MichaelOrlov committed:
Renames for shadowed callback_mutex_ from the EventHandlerBase

Note: We can't reuse `on_new_event_callback_mutex_` from the
EventHandlerBase because those mutex used to protect another type of
callbacks.

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov committed:
Temporary removes the on_new_[message]event on disable_callbacks()

Note: While we are temporary removes the on new message callback and all
 on new event callbacks from the underlying rmw layer to prevent them
 from being called. We can't guarantee that the currently executing
 on_new_[message]event callbacks are finished before we exit from the
 disable_callbacks().

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov commented:
@jmachowinski, kindly ping here to reiterate on the review of the last two commits.

MichaelOrlov committed:
Use recursive_mutex for callback lock in any_subscription_callback

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov commented:
Pulls: ros2/rclcpp#2985
Gist: https://gist.githubusercontent.com/MichaelOrlov/858df3b489887929e03235892f22539a/raw/905ba8c783b44ae6c3ff1615db00916120eb0d5b/ros2.repos
BUILD args:  --packages-above-and-dependencies rclcpp
TEST args:  --packages-above rclcpp
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/17604
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=26304)](http://ci.ros2.org/job/ci_linux/26304/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=19952)](http://ci.ros2.org/job/ci_linux-aarch64/19952/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=6001)](http://ci.ros2.org/job/ci_linux-rhel/6001/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=26046)](http://ci.ros2.org/job/ci_windows/26046/)

MichaelOrlov commented:
The CI is yellowish, but it seems it is unrelated.  
Does anybody know if this is okay and expected to be this way?

MichaelOrlov committed:
Address CI warnings about usage of deprecated rclcpp::spin_some(node)

- Use spin_some() from the rclcpp::executors::SingleThreadedExecutor
 directly.

Signed-off-by: Michael Orlov <morlovmr@gmail.com>

MichaelOrlov commented:
Ahh, I've figured it out. The CI is yellowish due to the new warnings 
> ‘void rclcpp::spin_some(Node::SharedPtr)’ is deprecated: use SingleThreadedExecutor::spin_some instead [-Wdeprecated-declarations] 346 | rclcpp::spin_some(node_); 

That is because I was using a deprecated `rclcpp::spin_some(node_)` in the newly added tests.

I've just fixed it with the new commit by using spin_some() from the rclcpp::executors::SingleThreadedExecutor
directly.

MichaelOrlov commented:
Re-run CI jobs that had warnings
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=26306)](http://ci.ros2.org/job/ci_linux/26306/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=19954)](http://ci.ros2.org/job/ci_linux-aarch64/19954/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=6004)](http://ci.ros2.org/job/ci_linux-rhel/6004/)

MichaelOrlov commented:
@ros-pull-request-builder retest this please

MichaelOrlov commented:
Re-run Linux CI job to see if it will pass after addressing pub-sub tests failure
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=26310)](http://ci.ros2.org/job/ci_linux/26310/)

MichaelOrlov commented:
The Linux CI job passed successfully. Re-running another two

* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=19974)](http://ci.ros2.org/job/ci_linux-aarch64/19974/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=6023)](http://ci.ros2.org/job/ci_linux-rhel/6023/)

MichaelOrlov commented:
All CI jobs passed green. Merging then.
